libavutil/hwcontext_mediafoundation.c

```diff
@@ -0,0 +1,445 @@
+                                NULL,
+                                NULL);
+        if (FAILED(hr)) {
+            av_log(ctx, loglevel, "failed to create D3D device\n");
+            goto error;
+        }
+
+        hr = IMFMediaBuffer_QueryInterface(d3d11_device, &IID_ID3D10Multithread, (void **)&multithread);
+        if (FAILED(hr)) {
+            av_log(ctx, loglevel, "could not get ID3D10Multithread\n");
+            goto error;
+        }
+
+        hr = ID3D10Multithread_SetMultithreadProtected(multithread, TRUE);
+        if (FAILED(hr)) {
+            av_log(ctx, loglevel, "failed to call ID3D10Multithread::SetMultithreadProtected\n");
+            ID3D10Multithread_Release(multithread);
+            goto error;
+        }
+
+        ID3D10Multithread_Release(multithread);
+    }
+
+    // If this code is enabled, we already link against this DLL.
+    // MFCreateDXGIDeviceManager is just not available on Windows 7.
+    mfplat_dll = GetModuleHandleW(L"mfplat.dll");
+    if (!mfplat_dll) {
+        av_log(ctx, loglevel, "mfplat.dll not present\n");
+        goto error;
+    }
+    pMFCreateDXGIDeviceManager = (void *)GetProcAddress(mfplat_dll, "MFCreateDXGIDeviceManager");
+    if (!pMFCreateDXGIDeviceManager) {
+        av_log(ctx, loglevel, "MFCreateDXGIDeviceManager not found\n");
+        goto error;
+    }
+
+    hr = pMFCreateDXGIDeviceManager(&token, &hwctx->d3d11_manager);
+    if (FAILED(hr)) {
+        av_log(ctx, loglevel, "failed to create IMFDXGIDeviceManager\n");
+        goto error;
+    }
+
+    hr = IMFDXGIDeviceManager_ResetDevice(hwctx->d3d11_manager, (IUnknown *)d3d11_device, token);
+    if (FAILED(hr)) {
+        av_log(ctx, loglevel, "failed to init IMFDXGIDeviceManager\n");
+        goto error;
+    }
+
+    ID3D11Device_Release(d3d11_device);
+    return 0;
+
+error:
+    if (d3d11_device)
+        ID3D11Device_Release(d3d11_device);
+    mf_uninit_d3d(ctx);
+    return AVERROR_EXTERNAL;
+}
+
+static int mf_create_d3d9_device(AVHWDeviceContext *ctx, int loglevel)
+{
+    AVMFDeviceContext *hwctx = ctx->hwctx;
+    MFDeviceContext *priv = ctx->internal->priv;
+    pDirect3DCreate9      *createD3D = NULL;
+    pCreateDeviceManager9 *createDeviceManager = NULL;
+    IDirect3DDevice9      *d3d9device = NULL;
+    HRESULT hr;
+    D3DPRESENT_PARAMETERS d3dpp = {0};
+    D3DDISPLAYMODE        d3ddm;
+    unsigned resetToken = 0;
+
+    if (hwctx->init_d3d9_device) {
+        d3d9device = hwctx->init_d3d9_device;
+        IDirect3DDevice9_AddRef(d3d9device);
+    } else {
+        priv->d3d9_dll = LoadLibraryW(L"d3d9.dll");
+        if (!priv->d3d9_dll) {
+            av_log(ctx, loglevel, "Failed to load D3D9 library\n");
+            goto fail;
+        }
+
+        createD3D = (pDirect3DCreate9 *)GetProcAddress(priv->d3d9_dll, "Direct3DCreate9");
+        if (!createD3D) {
+            av_log(ctx, loglevel, "Failed to locate Direct3DCreate9\n");
+            goto fail;
+        }
```

Comment:
Could we use Direct3DCreate9Ex instead of Direct3DCreate9 like in DXVA2?
DXVA2 code try to do both.
https://ffmpeg.org/pipermail/ffmpeg-devel/2017-February/206971.html
https://github.com/gopro/FFmpeg/blob/eacfcbae690f914a4b1b4ad06999f138540cc3d8/libavutil/hwcontext_dxva2.c#L549

Reply:
I implemented this in the next commit.  Please take a look.